docs/proposal-001_add-ons.md

```diff
@@ -0,0 +1,60 @@
+# Desing Proposal #001: Add-ons
+
+## What's an add-ons?
+
+An add-on extends functionality of a Kubernetes cluster, it may consist of a workload and/or configuration within the give cluster or the cloud provider. The workload, if present, would may classify as an operator or a controller, however that is not neccessary.
+
+The difference between add-ons and anything user decided to run themselves, is that from user's point of view, add-on is something they don't need to maintain directly. A cluster provider or bootstrap tool (i.e. eksctl), is expected to ensure ease of use, by providing minumum viable (that configuration to ensure end-to-end integration and compatibility), as well as catering for reconfiguration and upgrades.
+
+## Design Requirements and Goals
+
+Add-on must expose all of the following attributes:
+
+- name (fixed) *[TBD]*
+- version tag or revision (has default value that user can override)
+- source URL (has default value that user can override)
+- set of customisations (no default value, user defined)
+
+Internally, an add-on may consist of any and/or all of the following elements:
+
+- Kubernetes workload definitions
+- Kubernetes configuration objects - PersistentVolumes, ConfigMaps, ...etc
+- cloud provider resource definitons (e.g. CloudFormation sub-resources)
+- dependencies (supported Kubernetes versions and corresponding cloud providers, list of references to other add-ons)
+- incompatibilities (supported Kubernetes versions and corresponding cloud providers, list of reference to other add-ons)
+
+In other words, it's ought to be possible to define an add-on that only includes:
+
+- dependencies on other add-ons
+- cloud provider resource definitions
+- Kubernetes configuration objects
+- Kubernetes workload definitions
+
+The implementation must also ensure:
+
+- fully deterministic behaviour for a given version of eksctl, thereby it should have
+    - strict versioning of built-in add-ons by default
+    - any externally-sourced config has to be pinned down or vendored *[TBD]*
+    - there should be an option to use latest version or altenative source URL
+- user should be able:
+    - remove any add-on (including it's dependencies) at any time
+    - customise an add-on in an arbitrary fashion
+    - upgrade an add-on
+    - install add-on at any time
+- an add-on may consists exclusively of
+    - a reference to any Helm chart (or a copy of one)
```

Comment:
I would consider `helm` itself an addon, do you think we should allow a chart as a packaging format?

Reply:
An addon could be packaged as a Helm chart but to apply it you don't necessary need Tiller running on the server. eksctl could detect if Tiller is not running in the cluster and do `helm template | kubectl apply`

Reply:
Just my two cents reading this: I would try to add additional options of templating as not everybody likes helm or wants to have it as a dependency. This doesn't need to be in a first implementation, but probably it makes sense to make it in the proposal if you agree.

Reply:
Yes, this proposal in the current form assumes there would a few different options. Helm is a popular choice of course, especially due to the fact that there are so many existing community charts.

Reply:
@christopherhein I would use a yaml file to provide the list of addons and the setting params for each addon.

eksctl -f ./config.yaml

```yaml
apiVersion: eksctl.io/v1alpha1
kind: Config
metadata:
  name: eks-dev
spec:
  addons:
    - name: cert-manager
      chart:
        repository: https://kubernetes-charts.storage.googleapis.com
        name: cert-manager
        version: v0.5.0
      values:
        createCustomResource: true
        rbac:
          create: true
    - name: kured
      chart:
        repository: https://kubernetes-charts.storage.googleapis.com
        name: kured
        version: 0.1.0
      values:
        updateStrategy: OnDelete
        rbac:
          create: true
```

Reply:
I love it, sounds like a great way to handle this.

Reply:
Thanks, Stefan! For now, I'd have distinct add-on objects, as I'd rather leave cluster object undefined for until we get to implementing Cluster API (right now the plan is to implement add-ons first). Ultimately, add-on objects can be inlined in cluster object at a later point. I've added some examples just now, will incorporate what you shown as well.

Reply:
Multiple templating/packing options would be great (helm being one obviously). I like the format that @stefanprodan listed but personally i'd prefer separate add-on resources. This would then allow our teams to each look after the addons they are interested in having (ensuring idempotency).

Reply:
I really like this too. +1 for trying to bring this into the Cluster object ASAP though, to avoid forking.